test/common/http/filter/cors_filter_test.cc

```diff
@@ -0,0 +1,251 @@
+#include <chrono>
+#include <memory>
+
+#include "envoy/event/dispatcher.h"
+
+#include "common/http/filter/cors_filter.h"
+#include "common/http/header_map_impl.h"
+#include "common/stats/stats_impl.h"
+
+#include "test/mocks/buffer/mocks.h"
+#include "test/mocks/http/mocks.h"
+#include "test/test_common/printers.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+using testing::DoAll;
+using testing::InSequence;
+using testing::Invoke;
+using testing::NiceMock;
+using testing::Return;
+using testing::ReturnRef;
+using testing::SaveArg;
+using testing::_;
+
+namespace Http {
+
+class CorsFilterTest : public testing::Test {
+public:
+  CorsFilterTest() : config_{new CorsFilterConfig{}}, filter_(config_) {
+    cors_policy_.reset(new Router::TestCorsPolicy());
+    cors_policy_->enabled_ = true;
+    cors_policy_->allow_origin_ = "*";
+    cors_policy_->allow_methods_ = "GET";
+    cors_policy_->allow_headers_ = "content-type";
+    cors_policy_->expose_headers_ = "content-type";
+    cors_policy_->allow_credentials_ = false;
+    cors_policy_->max_age_ = "0";
+
+    ON_CALL(decoder_callbacks_.route_->route_entry_, corsPolicy())
+        .WillByDefault(ReturnRef(*cors_policy_));
+
+    ON_CALL(decoder_callbacks_.route_->route_entry_.virtual_host_, corsPolicy())
+        .WillByDefault(ReturnRef(*cors_policy_));
+
+    filter_.setDecoderFilterCallbacks(decoder_callbacks_);
+    filter_.setEncoderFilterCallbacks(encoder_callbacks_);
+  }
+
+  NiceMock<MockStreamDecoderFilterCallbacks> decoder_callbacks_;
+  NiceMock<MockStreamEncoderFilterCallbacks> encoder_callbacks_;
+  std::shared_ptr<CorsFilterConfig> config_;
+  CorsFilter filter_;
+  Buffer::OwnedImpl data_;
+  TestHeaderMapImpl request_headers_;
+  std::shared_ptr<Router::TestCorsPolicy> cors_policy_;
+};
+
+TEST_F(CorsFilterTest, NonOptionsRequest) {
+  Http::TestHeaderMapImpl request_headers{{"method", "get"}};
+
+  EXPECT_EQ(FilterHeadersStatus::Continue, filter_.decodeHeaders(request_headers, false));
+  EXPECT_EQ(FilterDataStatus::Continue, filter_.decodeData(data_, false));
+  EXPECT_EQ(FilterTrailersStatus::Continue, filter_.decodeTrailers(request_headers_));
+
+  EXPECT_EQ(FilterHeadersStatus::Continue, filter_.encodeHeaders(request_headers_, false));
+  EXPECT_EQ(FilterDataStatus::Continue, filter_.encodeData(data_, false));
+  EXPECT_EQ(FilterTrailersStatus::Continue, filter_.encodeTrailers(request_headers_));
+}
+
+TEST_F(CorsFilterTest, OptionsRequestWithoutOrigin) {
+  Http::TestHeaderMapImpl request_headers{{"method", "options"}};
+
+  EXPECT_EQ(FilterHeadersStatus::Continue, filter_.decodeHeaders(request_headers, false));
+  // EXPECT_EQ(false, filter_.is_cors_request_);
```

Comment:
@alyssawilk is there any way to do this check? Maybe only by having a public accessor?

Reply:
Thinking about it i'll go with `friend class CorsFilterTest;` is that ok?

Reply:
Good question - I'm new enough I'm not entirely sure of the answer :-P   I'm fan of test peers and I'd be inclined to claim DnsResolverImplPeer is precedent (and I prefer having Test in the name) enough to say go for it, but @mattklein123 may have a different opinion :-)

Reply:
A friend class is fine if you really need it, however, in this case my preference would be to just test the outward behaviors. Meaning, you don't really care how the filter works, you just care that it's not handling in this case and returning continue. So that would be what I would do. I don't have a very strong preference though.

Reply:
I was doing that first but then I had the problem that in almost all cases it returns continue and I cannot test with confidence if it really went into the specific if case. I also think black box testing is ok in general. But since there is a spec for when to return I wanted to test that somehow. If it's just an implementation detail I wouldn't want to test it. 
I have some problems with the tests now. While testing in my dev environment works some tests fail. 

I never heard the term test peers and Google didn't bring anything up. But I guess it means friending the class? Judging from the name. :)

Reply:
Yes basically.